google-cloud-bigquery/lib/google/cloud/bigquery/model/aggregate_classification_metrics.rb

```diff
@@ -0,0 +1,98 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+module Google
+  module Cloud
+    module Bigquery
+      class Model
+        # Aggregate metrics for classification/classifier models. For
+        # multi-class models, the metrics are either macro-averaged or
+        # micro-averaged. When macro-averaged, the metrics are calculated for
+        # each label and then an unweighted average is taken of those values.
+        # When micro-averaged, the metric is calculated globally by counting the
+        # total number of correctly predicted rows.
+        class AggregateClassificationMetrics
```

Comment:
Are you generating these type definitions by hand or are you consuming something? Ex: python consumes the proto representations and just uses the json serialization functionality, java and python expose the discovery types directly with special caveats about volatility.  There's enough churn due to the ML backend change velocity that you'll want to keep out of the business of maintaining them by hand, as there's frequent edits here.

The advantage we have here is that these are output-only messages, since all resource creation is done via CREATE MODEL queries.

Reply:
These files were written by hand, but we can change this approach and generate off the discovery document.

We have so far been very intentional about not leaking the apiary/discovery layer, but for these particular values I am not opposed. I see a few options moving forward. Which would you prefer?

1) Maintain veneer versions of these structures. Changes to the structures in the discovery document will require changes to these files. This can be manual or automated, but will require changes to this gem for new data to be accessed.
2) Return Arrays and Hashes of values instead of documented objects. This can either be the apiary Hash representation, or the API's JSON structure. (`training_options` vs. `trainingOptions`) This means new data can be accessed when available.
3) Return the apiary/Google API Client objects directly, with a caveat that the version of google-api-client used will determine what data is returned. This also means new data can be accessed when available.

Reply:
Does ruby manually pin its discovery dependencies, or does it pick up new versions as they're available?  I want to avoid having anyone having to monitor discovery for changes, and propagating them by hand.

Is option 2 fairly idiomatic for ruby folks, or would they have a stronger preference towards the complex typing we'd get from converting messages to an appropriate object/type?  There's also the option of consuming the [proto message](https://github.com/googleapis/googleapis/blob/master/google/cloud/bigquery/v2/model.proto) to generate the type in similar fashion to python, but it depends on support for converting between proto types and json representation.

Reply:
> Does ruby manually pin its discovery dependencies, or does it pick up new versions as they're available?

The google-cloud-bigquery gem maintains a minimum dependency on google-api-client. So anything that the google-cloud-bigquery gem exposes needs to be present in the minimum google-api-client version. Newer google-api-client versions can be used, but they are assumed to be compatible with the minimum version specified. Adding new features in the google-cloud-bigquery gem will bump the dependency on the google-api-client gem to the lowest version that contains it.

If the API returns additional data then it is assumed that those changes are reflected in the apiary discovery document (and proto files). That will kick off a new release of the google-api-client gem that includes those features, which the google-cloud-bigquery gem would then set as the new minimum dependency when it exposes the data. This is most pronounced in option 1.

Option 2 is the most flexible, as it would allow the google-cloud-bigquery gem to expose the data for whatever version of google-api-client was being used, and also move away from google-api-client in the future. Option 3 cements the dependency on google-api-client, and makes it part of the google-cloud-bigquery public API.

Reply:
Sounds like option 2 is the best choice here, as it gives us the ability to keep up with backend changes and not encumber the veneer with the explicit dependency on the discovery library.

Reply:
One thing we can do is bypass the Google API Client object model, and parse the JSON directly. That would allow us to return all the data that the API returns regardless of which version of google-api-client is being used. Is that something that you would like us to do?

Reply:
I've updated the PR and changed `#training_runs` to return the deserialized JSON from the API, so any changes to the data returned will be visible immediately. PTAL.